modules/mastodon/index.js

```diff
@@ -0,0 +1,324 @@
+      });
+    });
+  });
+}
+
+function defaultConfig() {
+  return { instances: require('./default_instances.json') };
+}
+
+function writeConfig() {
+  bot.writeDataFile('config.json', JSON.stringify(moduleConfig), (err) => {
+    if (err) { log.error(`error writing mastodon module data: ${err}`); }
+  });
+}
+
+function checkHostname(hostname) {
+  if (moduleConfig.blocks && moduleConfig.blocks.includes(hostname)) { return false; }
+
+  return moduleConfig.instances && moduleConfig.instances.includes(hostname);
+}
+
+/* TODO: Add support for validating GNU social? */
+/* TODO: Don't check instances more than once a day. */
+function forceCheckHostname(hostname, fn) {
+  robotRequest(
+    { url: `https://${hostname}/api/v1/instance` },
+    (error, response, bodyRaw) => {
+      let body;
+
+      if (error || response.statusCode >= 400) { return fn(false); }
+
+      try {
+        body = JSON.parse(bodyRaw);
+      } catch (e) {
+        return fn(false);
+      }
+
+      if (!body.uri) { return fn(false); }
+
+      fn(true);
+      log.info(`adding ${response.request.host} to instance list`);
+      if (moduleConfig.instances) {
+        moduleConfig.instances.push(response.request.host);
+      } else {
+        moduleConfig.instances = [response.request.host];
+      }
+      return writeConfig();
+    },
+  );
+}
+
+/*
+ * fetchResource only attempts to fetch the URL given as a particular content
+ * type. This is less than what a full ActivityPub-implementing client needs to
+ * handle; following Link headers and <link> HTML tags of the correct content
+ * types would also be required.
+ */
+function fetchResource(url, fn) {
+  robotRequest({
+    url,
+    headers: { Accept: 'application/ld+json; profile="https://www.w3.org/ns/activitystreams", application/activity+json' },
+  }, (error, response, body) => {
+    if (error) {
+      fn(error);
+    } else if (response.statusCode >= 400) {
+      fn(new Error(`response status code is ${response.statusCode}`));
+    } else if (!['application/ld+json', 'application/activity+json'].includes(contentType.parse(response.headers['content-type']).type)) {
+      fn(new Error('resource is not an ActivityPub record'));
+    } else {
+      try {
+        const record = JSON.parse(body);
+        if (!['Note', 'Person'].includes(record.type)) {
+          fn(new Error('record is not a Note or Person'));
+        } else {
+          fn(null, record);
+        }
+      } catch (e) {
+        fn(e);
+      }
+    }
+  });
+}
+
+function htmlToText(s) {
+  return htmlToTextMod.fromString(s, {
+    wordwrap: null,
+    ignoreHref: true,
+  });
+}
+
+function formatRecord(record, fn) {
+  if (record.type === 'Note') {
+    fetchResource(record.attributedTo, (error, userRecord) => {
+      let str = '';
+      if (error) {
+        str += '(error getting user): ';
+      } else {
+        str += `${userRecord.name} (@${userRecord.preferredUsername}): `;
+      }
+      if (record.sensitive) {
```

Comment:
Future work, but I think that if invoked with the command it would make sense to override CWs and show the content. The person pasting the link to irc is then consciously deciding that it's suitable content for the channel.

Reply:
I disagree, from the perspective of our channel.

Given some of the content I sometimes see discussed, I don't think each user has an understanding of the content warnings folks need.